README.md
# material_dialog

A wrapper on top of alert dialog provided by flutter.

## Demo

<p>
  <img src="screenshots/Screenshot_20191204-153700.jpg" width="300">
  <img src="screenshots/Screenshot_20191204-153705.jpg" width="300">
  <img src="screenshots/Screenshot_20191204-153710.jpg" width="300">
  <img src="screenshots/Screenshot_20191204-153716.jpg" width="300">
  <img src="screenshots/Screenshot_20191204-153721.jpg" width="300">
</p>

Use this package as a library
-----------------------------

**1. Depend on it**

Add this to your package's pubspec.yaml file:

```
dependencies:
  material_dialog: ^0.0.x
```

**2. Install it**

You can install packages from the command line:

with Flutter

```
$ flutter packages get
```

Alternatively, your editor might support flutter packages get. Check the docs for your editor to learn more.

**3. Import it**

Now in your Dart code, you can use:

```
import 'package:material_dialog/material_dialog.dart';
```

How to use
----------

Using Material Dialog is quite easy, all you need to do is 

```dart
MaterialDialog(
  content: Text(_alertText),
  actions: <Widget>[
    FlatButton(
      child: Text(
        'CANCEL',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.agree);
      },
    ),
    FlatButton(
      child: Text(
        'OK',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.cancel);
      },
    ),
  ],
);
```

Use below method to show dialog

```dart
  void showDemoDialog<T>({BuildContext context, Widget child}) {
    showDialog<T>(
      context: context,
      builder: (BuildContext context) => child,
    ).then<void>((T value) {
      // The value passed to Navigator.pop() or null.
      if (value != null) {
        _scaffoldKey.currentState.showSnackBar(SnackBar(
          content: Text('You selected: $value'),
        ));
      }
    });
  }
```

## Simple Alert Dialog
```dart
Widget _buildAlertDialog(ThemeData theme, BuildContext context) {
return MaterialDialog(
  content: Text(_alertText),
  actions: <Widget>[
    FlatButton(
      child: Text(
        'CANCEL',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.agree);
      },
    ),
    FlatButton(
      child: Text(
        'OK',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.cancel);
      },
    ),
  ],
);
}
```

## Alert Dialog With Title
```dart
Widget _buildAlertDialogWithTitle(ThemeData theme, BuildContext context) {
return MaterialDialog(
  title: Text(_alertTitle),
  subTitle: Text('Subtitle'),
  content: Text(_alertText),
  actions: <Widget>[
    FlatButton(
      child: Text(
        'CANCEL',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.agree);
      },
    ),
    FlatButton(
      child: Text(
        'OK',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.cancel);
      },
    ),
  ],
);
}
```

## Alert Dialog With Close & Back Button
```dart
Widget _buildAlertDialogWithCloseAndBackButton(
  ThemeData theme, BuildContext context) {
return MaterialDialog(
  title: Text(_alertTitle),
  subTitle: Text('Subtitle'),
  content: Text(_alertText),
  enableBackButton: true,
  enableCloseButton: true,
  onBackButtonClicked: () {
    Navigator.pop(context, DialogDemoAction.agree);
  },
  onCloseButtonClicked: () {
    Navigator.pop(context, DialogDemoAction.cancel);
  },
  actions: <Widget>[
    FlatButton(
      child: Text(
        'CANCEL',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.agree);
      },
    ),
    FlatButton(
      child: Text(
        'OK',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.cancel);
      },
    ),
  ],
);
}
```

## Alert Dialog With Children
```dart
Widget _buildAlertDialogWithChildren(
  ThemeData theme, BuildContext context, bool isFullScreen) {
return MaterialDialog(
  borderRadius: 8.0,
  enableFullHeight: isFullScreen,
  enableFullWidth: isFullScreen,
  enableCloseButton: true,
  closeButtonColor: Colors.white,
  headerColor: Theme.of(context).primaryColor,
  title: Text(
    _alertTitle,
    style: TextStyle(
      color: Colors.white,
      fontSize: 18.0,
    ),
  ),
  subTitle: Text(
    'Subtitle',
    style: TextStyle(
      color: Colors.white70,
      fontSize: 12.0,
    ),
  ),
  onCloseButtonClicked: () {
    Navigator.pop(context, DialogDemoAction.cancel.toString());
  },
  children: <Widget>[
    Text(
      _alertTitle,
      style: TextStyle(
        fontSize: 18.0,
      ),
    ),
    SizedBox(height: 8.0),
    Text(
      _alertText,
      style: TextStyle(
        fontSize: 12.0,
      ),
    ),
    SizedBox(height: 16.0),
    TextField(
      decoration: InputDecoration(hintText: 'Enter Username'),
    ),
  ],
  actions: <Widget>[
    FlatButton(
      child: Text(
        'CANCEL',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.cancel.toString());
      },
    ),
    FlatButton(
      child: Text(
        'OK',
        style: Theme.of(context).textTheme.button.copyWith(
            fontSize: 12.0, color: Theme.of(context).primaryColor),
      ),
      onPressed: () {
        Navigator.pop(context, DialogDemoAction.agree.toString());
      },
    )
  ],
);
}
```

List of params
--------------

| Property              | Type         | Default Value | Description                                                                                                                                       |
|-----------------------|--------------|---------------|---------------------------------------------------------------------------------------------------------------------------------------------------|
| title                 | Widget       | null          | The (optional) title of the dialog is displayed in a large font at the top of the dialog.                                                         |
| subTitle              | Widget       | null          | The (optional) subtitle of the dialog is displayed below title of the dialog.                                                                     |
| content               | Widget       | null          | The (optional) content of the dialog is displayed in the center of the dialog in a lighter font.                                                  |
| actions               | List<Widget> | null          | The (optional) set of actions that are displayed at the bottom of the dialog.                                                                     |
| children              | List<Widget> | null          | The (optional) content of the dialog is displayed in a [SingleChildScrollView] underneath the title.                                              |
| headerColor           | Color        | null          | The (optional) header color is displayed in the header background.                                                                                |
| backButtonColor       | Color        | Colors.white  | The (optional) back button color. By default its set to white.                                                                                    |
| closeButtonColor      | Color        | Colors.white  | The (optional) close button color. By default its set to white.                                                                                   |
| onBackButtonClicked   | VoidCallback | null          | A callback function to get back button event from dialog. If back button is enabled this callback has to be provided in-order to get callbacks.   |
| onCloseButtonClicked  | VoidCallback | null          | A callback function to get close button event from dialog. If close button is enabled this callback has to be provided in-order to get callbacks. |
| enableBackButton      | bool         | false         | The (optional) value to enable/disable back button for a dialog.                                                                                  |
| enableCloseButton     | bool         | false         | The (optional) value to enable/disable close button for a dialog.                                                                                  |
| enableFullWidth       | bool         | true          | The (optional) value to stretch dialog to its max width.                                                                                          |
| enableFullHeight      | bool         | false         | The (optional) value to stretch dialog to its max height.                                                                                         |
| borderRadius          | double       | 10.0          | The (optional) border radius of a dialog. by default its 10.0.                                                                                    |
   
Thanks
------
If you liked my work, don’t forget to ⭐ star the repo to show your support.
                  

CHANGELOG.md
## [0.0.1] - 11-Sep-2019.
* Released package.

## [0.0.2] - 11-Sep-2019.
* added custom layout
* added header color
* added back button and listener
* added close button and listener
* added border radius
* added enable/disable back and close button

## [0.0.3] - 12-Sep-2019.

* added divider for actions
## [0.0.4] - 12-Sep-2019.

* added readme.md file

## [0.0.5] - 12-Sep-2019.
* updated onBackClick -> onBackButtonClicked
* updated onCloseClick -> onCloseButtonClicked
* added check for empty space in portrait mode

## [0.0.6] - 16-Sep-2019.
* fixed title padding 

## [0.0.7] - 26-Sep-2019.
* added support for background color

## [0.0.8] - 04-Dec-2019.
* fixed minor UI issues
* refactored examples

## [0.0.9] - 04-Dec-2019.
* updated readme.md file



lib/material_dialog.dart
library material_dialog;

import 'package:flutter/material.dart';
import 'package:material_dialog/widgets/dialog.dart' as dialog;

class MaterialDialog extends StatelessWidget {
  final Widget title;
  final Widget subTitle;
  final Widget content;
  final List<Widget> actions;
  final List<Widget> children;
  final Color headerColor;
  final Color backButtonColor;
  final Color closeButtonColor;
  final Color backgroundColor;
  final bool enableFullWidth;
  final bool enableFullHeight;
  final bool enableBackButton;
  final bool enableCloseButton;
  final double borderRadius;
  final VoidCallback onBackButtonClicked;
  final VoidCallback onCloseButtonClicked;

  const MaterialDialog({
    Key key,
    this.title,
    this.subTitle,
    this.content,
    this.actions,
    this.children,
    this.enableFullWidth = false,
    this.enableFullHeight = false,
    this.headerColor,
    this.backButtonColor = Colors.black,
    this.closeButtonColor = Colors.black,
    this.borderRadius = 10.0,
    this.onBackButtonClicked,
    this.onCloseButtonClicked,
    this.enableBackButton = false,
    this.enableCloseButton = false,
    this.backgroundColor = Colors.white,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return dialog.SimpleDialog(
      backgroundColor: backgroundColor,
      shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(borderRadius)),
      title: title ?? null,
      subTitle: subTitle ?? null,
      content: content,
      actions: actions,
      children: children,
      headerColor: headerColor,
      backButtonColor: backButtonColor,
      closeButtonColor: closeButtonColor,
      borderRadius: borderRadius,
      onBackClick: onBackButtonClicked,
      onCloseClick: onCloseButtonClicked,
      enableBackButton: enableBackButton,
      enableCloseButton: enableCloseButton,
      enableFullWidth: enableFullWidth,
      enableFullHeight: enableFullHeight,
    );
  }
}


lib/widgets/button_bar.dart
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

/// An end-aligned row of buttons.
///
/// Places the buttons horizontally according to the padding in the current
/// [ButtonTheme]. The children are laid out in a [Row] with
/// [MainAxisAlignment.end]. When the [Directionality] is [TextDirection.ltr],
/// the button bar's children are right justified and the last child becomes
/// the rightmost child. When the [Directionality] [TextDirection.rtl] the
/// children are left justified and the last child becomes the leftmost child.
///
/// Used by [Dialog] to arrange the actions at the bottom of the dialog.
///
/// See also:
///
///  * [RaisedButton], a kind of button.
///  * [FlatButton], another kind of button.
///  * [Card], at the bottom of which it is common to place a [ButtonBar].
///  * [Dialog], which uses a [ButtonBar] for its actions.
///  * [ButtonTheme], which configures the [ButtonBar].
class ButtonBar extends StatelessWidget {
  /// Creates a button bar.
  ///
  /// The alignment argument defaults to [MainAxisAlignment.end].
  const ButtonBar({
    Key key,
    this.alignment = MainAxisAlignment.end,
    this.mainAxisSize = MainAxisSize.max,
    this.children = const <Widget>[],
  }) : super(key: key);

  /// How the children should be placed along the horizontal axis.
  final MainAxisAlignment alignment;

  /// How much horizontal space is available. See [Row.mainAxisSize].
  final MainAxisSize mainAxisSize;

  /// The buttons to arrange horizontally.
  ///
  /// Typically [RaisedButton] or [FlatButton] widgets.
  final List<Widget> children;

  @override
  Widget build(BuildContext context) {
    final ButtonThemeData buttonTheme = ButtonTheme.of(context);
    // We divide by 4.0 because we want half of the average of the left and right padding.
    final double paddingUnit = buttonTheme.padding.horizontal / 4.0;
    final Widget child = Row(
      mainAxisAlignment: alignment,
      mainAxisSize: mainAxisSize,
      children: children.map<Widget>((Widget child) {
        return Padding(
          padding: EdgeInsets.symmetric(horizontal: paddingUnit),
          child: child,
        );
      }).toList(),
    );
    switch (buttonTheme.layoutBehavior) {
      case ButtonBarLayoutBehavior.padded:
        return Padding(
          padding: EdgeInsets.symmetric(
//            vertical: paddingUnit,
//            horizontal: paddingUnit,
              ),
          child: child,
        );
      case ButtonBarLayoutBehavior.constrained:
        return Container(
          padding: EdgeInsets.symmetric(horizontal: paddingUnit),
          constraints: const BoxConstraints(minHeight: 52.0),
          alignment: Alignment.center,
          child: child,
        );
    }
    assert(false);
    return null;
  }
}


example/lib/ui/demo/dialog_demo.dart
import 'package:example/ui/demo/dialog_action.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:material_dialog/material_dialog.dart';

class DialogDemo extends StatefulWidget {
  @override
  DialogDemoState createState() => DialogDemoState();
}

class DialogDemoState extends State<DialogDemo> {
  // global key
  final GlobalKey<ScaffoldState> _scaffoldKey = GlobalKey<ScaffoldState>();

  // title
  final String _alertTitle = 'Discard draft?';

  // message
  final String _alertText =
      'Let Google help apps determine location. This means sending anonymous location '
      'data to Google, even when no apps are running.';

  @override
  Widget build(BuildContext context) {
    final ThemeData theme = Theme.of(context);

    return Scaffold(
      key: _scaffoldKey,
      appBar: AppBar(
        title: const Text('Custom Material Dialogs'),
      ),
      body: ListView(
        padding: const EdgeInsets.symmetric(vertical: 24.0, horizontal: 72.0),
        children: <Widget>[
          RaisedButton(
            child: const Text('ALERT'),
            onPressed: () {
              showDemoDialog<DialogDemoAction>(
                context: context,
                child: _buildAlertDialog(theme, context),
              );
            },
          ),
          RaisedButton(
            child: const Text('ALERT WITH TITLE'),
            onPressed: () {
              showDemoDialog<DialogDemoAction>(
                context: context,
                child: _buildAlertDialogWithTitle(theme, context),
              );
            },
          ),
          RaisedButton(
            child: const Text(
              'ALERT WITH CLOSE AND BACK BUTTON',
              textAlign: TextAlign.center,
            ),
            onPressed: () {
              showDemoDialog<DialogDemoAction>(
                context: context,
                child: _buildAlertDialogWithCloseAndBackButton(theme, context),
              );
            },
          ),
          RaisedButton(
            child: const Text('ALERT WITH CHILDREN'),
            onPressed: () {
              showDemoDialog<String>(
                context: context,
                child: _buildAlertDialogWithChildren(theme, context, false),
              );
            },
          ),
          RaisedButton(
            child: const Text('FULLSCREEN'),
            onPressed: () {
              showDemoDialog<String>(
                context: context,
                child: _buildAlertDialogWithChildren(theme, context, true),
              );
            },
          ),
        ]
            // Add a little space between the buttons
            .map<Widget>((Widget button) {
          return Container(
            padding: const EdgeInsets.symmetric(vertical: 8.0),
            child: button,
          );
        }).toList(),
      ),
    );
  }

  // Example 1: Basic Alert Dialog ---------------------------------------------
  Widget _buildAlertDialog(ThemeData theme, BuildContext context) {
    return MaterialDialog(
      content: Text(_alertText),
      actions: <Widget>[
        FlatButton(
          child: Text(
            'CANCEL',
            style: Theme.of(context).textTheme.button.copyWith(
                fontSize: 12.0, color: Theme.of(context).primaryColor),
          ),
          onPressed: () {
            Navigator.pop(context, DialogDemoAction.agree);
          },
        ),
        FlatButton(
          child: Text(
            'OK',
            style: Theme.of(context).textTheme.button.copyWith(
                fontSize: 12.0, color: Theme.of(context).primaryColor),
          ),
          onPressed: () {
            Navigator.pop(context, DialogDemoAction.cancel);
          },
        ),
      ],
    );
  }

  // Example 2: Basic Alert Dialog And Title -----------------------------------
  Widget _buildAlertDialogWithTitle(ThemeData theme, BuildContext context) {
    return MaterialDialog(
      title: Text(_alertTitle),
      subTitle: Text('Subtitle'),
      content: Text(_alertText),
      actions: <Widget>[
        FlatButton(
          child: Text(
            'CANCEL',
            style: Theme.of(context).textTheme.button.copyWith(
                fontSize: 12.0, color: Theme.of(context).primaryColor),
          ),
          onPressed: () {
            Navigator.pop(context, DialogDemoAction.agree);
          },
        ),
        FlatButton(
          child: Text(
            'OK',
            style: Theme.of(context).textTheme.button.copyWith(
                fontSize: 12.0, color: Theme.of(context).primaryColor),
          ),
          onPressed: () {
            Navigator.pop(context, DialogDemoAction.cancel);
          },
        ),
      ],
    );
  }

  // Example 3: Basic Alert Dialog With Close & Back Buttons -------------------
  Widget _buildAlertDialogWithCloseAndBackButton(
      ThemeData theme, BuildContext context) {
    return MaterialDialog(
      title: Text(_alertTitle),
      subTitle: Text('Subtitle'),
      content: Text(_alertText),
      enableBackButton: true,
      enableCloseButton: true,
      onBackButtonClicked: () {
        Navigator.pop(context, DialogDemoAction.agree);
      },
      onCloseButtonClicked: () {
        Navigator.pop(context, DialogDemoAction.cancel);
      },
      actions: <Widget>[
        FlatButton(
          child: Text(
            'CANCEL',
            style: Theme.of(context).textTheme.button.copyWith(
                fontSize: 12.0, color: Theme.of(context).primaryColor),
          ),
          onPressed: () {
            Navigator.pop(context, DialogDemoAction.agree);
          },
        ),
        FlatButton(
          child: Text(
            'OK',
            style: Theme.of(context).textTheme.button.copyWith(
                fontSize: 12.0, color: Theme.of(context).primaryColor),
          ),
          onPressed: () {
            Navigator.pop(context, DialogDemoAction.cancel);
          },
        ),
      ],
    );
  }

  // Example 4: Basic Alert Dialog With Children -------------------------------
  Widget _buildAlertDialogWithChildren(
      ThemeData theme, BuildContext context, bool isFullScreen) {
    return MaterialDialog(
      borderRadius: 8.0,
      enableFullHeight: isFullScreen,
      enableFullWidth: isFullScreen,
      enableCloseButton: true,
      closeButtonColor: Colors.white,
      headerColor: Theme.of(context).primaryColor,
      title: Text(
        _alertTitle,
        style: TextStyle(
          color: Colors.white,
          fontSize: 18.0,
        ),
      ),
      subTitle: Text(
        'Subtitle',
        style: TextStyle(
          color: Colors.white70,
          fontSize: 12.0,
        ),
      ),
      onCloseButtonClicked: () {
        Navigator.pop(context, DialogDemoAction.cancel.toString());
      },
      children: <Widget>[
        Text(
          _alertTitle,
          style: TextStyle(
            fontSize: 18.0,
          ),
        ),
        SizedBox(height: 8.0),
        Text(
          _alertText,
          style: TextStyle(
            fontSize: 12.0,
          ),
        ),
        SizedBox(height: 16.0),
        TextField(
          decoration: InputDecoration(hintText: 'Enter Username'),
        ),
      ],
      actions: <Widget>[
        FlatButton(
          child: Text(
            'CANCEL',
            style: Theme.of(context).textTheme.button.copyWith(
                fontSize: 12.0, color: Theme.of(context).primaryColor),
          ),
          onPressed: () {
            Navigator.pop(context, DialogDemoAction.cancel.toString());
          },
        ),
        FlatButton(
          child: Text(
            'OK',
            style: Theme.of(context).textTheme.button.copyWith(
                fontSize: 12.0, color: Theme.of(context).primaryColor),
          ),
          onPressed: () {
            Navigator.pop(context, DialogDemoAction.agree.toString());
          },
        )
      ],
    );
  }

  // general methods:-----------------------------------------------------------
  void showDemoDialog<T>({BuildContext context, Widget child}) {
    showDialog<T>(
      context: context,
      builder: (BuildContext context) => child,
    ).then<void>((T value) {
      // The value passed to Navigator.pop() or null.
      if (value != null) {
        _scaffoldKey.currentState.showSnackBar(SnackBar(
          content: Text('You selected: $value'),
        ));
      }
    });
  }
}
